// recyclerlistview/dist/web/platform/web/viewrenderer/ViewRenderer.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var BaseViewRenderer_1 = require("../../../core/viewrenderer/BaseViewRenderer");
/***
 * View renderer is responsible for creating a container of size provided by LayoutProvider and render content inside it.
 * Also enforces a logic to prevent re renders. RecyclerListView keeps moving these ViewRendereres around using transforms to enable recycling.
 * View renderer will only update if its position, dimensions or given data changes. Make sure to have a relevant shouldComponentUpdate as well.
 * This is second of the two things recycler works on. Implemented both for web and react native.
 */
var ViewRenderer = /** @class */ (function (_super) {
    __extends(ViewRenderer, _super);
    function ViewRenderer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._dim = { width: 0, height: 0 };
        _this._isFirstLayoutDone = false;
        return _this;
    }
    ViewRenderer.prototype.componentDidMount = function () {
        this._checkSizeChange();
    };
    ViewRenderer.prototype.componentDidUpdate = function () {
        this._checkSizeChange();
    };
    ViewRenderer.prototype.render = function () {
        var _this = this;
        var styleObj = this.props.forceNonDeterministicRendering
            ? {
                WebkitTransform: this._getTransform(),
                left: 0,
                opacity: this._isFirstLayoutDone ? 1 : 0,
                position: "absolute",
                top: 0,
                transform: this._getTransform(),
            }
            : {
                WebkitTransform: this._getTransform(),
                height: this.props.height,
                left: 0,
                overflow: "hidden",
                position: "absolute",
                top: 0,
                transform: this._getTransform(),
                width: this.props.width,
            };
        return (React.createElement("div", { ref: function (div) { return _this._mainDiv = div; }, style: styleObj }, this.renderChild()));
    };
    ViewRenderer.prototype._getTransform = function () {
        return "translate(" + this.props.x + "px," + this.props.y + "px)";
    };
    ViewRenderer.prototype._checkSizeChange = function () {
        if (this.props.forceNonDeterministicRendering && this.props.onSizeChanged) {
            var mainDiv = this._mainDiv;
            if (mainDiv) {
                this._dim.width = mainDiv.clientWidth;
                this._dim.height = mainDiv.clientHeight;
                if (this.props.width !== this._dim.width || this.props.height !== this._dim.height) {
                    this.props.onSizeChanged(this._dim, this.props.index);
                }
                else if (!this._isFirstLayoutDone) {
                    this._isFirstLayoutDone = true;
                    this.forceUpdate();
                }
            }
        }
        this._isFirstLayoutDone = true;
    };
    return ViewRenderer;
}(BaseViewRenderer_1.default));
exports.default = ViewRenderer;
